File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String strN = br.readLine();
        String strA = br.readLine();

        int N = Integer.parseInt(strN);
        int[] A = Arrays.stream(strA.split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int[] memo = new int[A.length];
        int pointer = 0; // memo pointer

        for (int el: A) {
            int low = 0;
            int high = pointer - 1;

            while (low <= high) {
                int mid = (low + high) / 2;

                if (memo[mid] < el) low = mid + 1;
                else high = mid - 1;
            }

            if (low >= pointer) {
                memo[pointer] = el;
                pointer++;
            } else {
                memo[low] = el;
            }
        }

        System.out.println(pointer);
        br.close();
    }
}

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.kt
fun main() {
    val strN: String? = readLine()
    val strA: String? = readLine()

    val N: Int = strN?.toInt() ?: return
    val A: List<Int> = strA?.split(" ")?.map { s -> s.toInt() } ?: return

    val memo: IntArray = IntArray(A.size) {0}
    var pointer: Int = 0

    for (el in A) {
        var low: Int = 0
        var high: Int = pointer - 1

        while (low <= high) {
            val mid: Int = (low + high) / 2

            if (memo[mid] < el) low = mid + 1
            else high = mid - 1
        }

        if (low >= pointer) {
            memo[pointer] = el
            pointer++
        } else {
            memo[low] = el
        }
    }

    println(pointer)
}